File: solutions/cpp/FirstBadVersion/FirstBadVersion.cpp
// Forward declaration of isBadVersion API.
bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int s = 1, e = n, m = 0;
        while (true) {
            if (s == e) {
                return s;
            } 
            m = (e - s) / 2 + s;
            if (isBadVersion(m)) {
                e = m;
            } else {
                s = m + 1;
            }
        }
    }
};

File: solutions/java/FirstBadVersion/FirstBadVersion.java
/* The isBadVersion API is defined in the parent class VersionControl.
      boolean isBadVersion(int version); */

public class Solution extends VersionControl {
    public int firstBadVersion(int n) {
        int start = 1;
        int end = n;
        int middle = 0;
        while (start < end) {
            middle = (end - start) / 2 + start;
            if (isBadVersion(middle)) {
                end = middle;
            } else {
                start = middle + 1;
            }
        }
        return start;
    }
}